server/server.go

```diff
@@ -3,52 +3,214 @@ package server
 	"io"
 	"log"
 	"mime/multipart"
 	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
 	"sync"
 )
 
-// Parses the client request and uploads the file.
+type assignmentTestingInformation struct {
+	CommandToExecute string
+	CommandToCompile string
+	WorkDir          string
+	RootDir          string
+	Output           string
+	CmdlineArgs      map[string]string
+}
+
+var assignTestingInfo assignmentTestingInformation
+
+func getSupportedLanguage(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	language := os.Getenv(environment.SupportedLanguage)
+
+	response, err := json.Marshal(language)
+	if err != nil {
+		log.Println(err)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(response)
+
+}
+
+// upload parses the client request and uploads the file.
 func upload(w http.ResponseWriter, r *http.Request) {
-	fmt.Println("File Upload Endpoint Hit")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	output := readFormData(r)
+
+	if len(output) <= 0 {
+		output += `"Upload Status":"Successfully Uploaded File(s)"`
+	}
+
+	response, err := json.Marshal(output)
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Write the response to be sent to client.
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(response)
+}
+
+// buildRun builds and runs the assignment uploaded.
+func build(w http.ResponseWriter, r *http.Request) {
 
 	w.Header().Set("Access-Control-Allow-Origin", "*")
 
-	responseString := readFormData(r)
+	var outputString string
+	var currDir string
+	var err error
+
+	// Navigate to the assignment working directory.
+	outputString, currDir = navigateToWorkDir()
+	if outputString == "" {
+		// Execute the compile command.
+		outputString, err = runCommand(assignTestingInfo.CommandToCompile)
+		if err != nil {
+			log.Println("error while building the assignment", err)
+			outputString = err.Error()
+		}
 
-	if len(responseString) <= 0 {
-		responseString += `"Upload Status":"Successfully Uploaded File(s)"`
+		// Navigate back to the code-runner working directory after successful execution.
+		err = os.Chdir(currDir)
+		if err != nil {
+			log.Println("error while navigating to the current directory", err)
+		}
 	}
 
-	responseS, err := json.Marshal(responseString)
+	if err == nil {
+		outputString = "Compiled successfully"
+	}
+	response, err := json.Marshal(outputString)
 	if err != nil {
 		log.Println(err)
 	}
+	// Write the response to be sent to client.
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(response)
+}
+
+// buildRun builds and runs the assignment uploaded.
+func run(w http.ResponseWriter, r *http.Request) {
+
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+
+	var outputString string
+	var currDir string
+	var err error
 
+	// Navigate to the assignment working directory.
+	outputString, currDir = navigateToWorkDir()
```

Comment:
from what I see, `outputString` stores the error message if any.
If yes, then it should be named `errorMsg`.

This brings to wonder why `navigateToWorkDir(...)` returns the error message and not the error itself.